v2/internal/reconcilers/arm/arm_client.go

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package arm
+
+import (
+	"context"
+	"reflect"
+	"sync"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/cloud"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-service-operator/v2/internal/config"
+	"github.com/Azure/azure-service-operator/v2/internal/genericarmclient"
+	"github.com/Azure/azure-service-operator/v2/internal/metrics"
+	"github.com/Azure/azure-service-operator/v2/internal/resolver"
+	"github.com/Azure/azure-service-operator/v2/internal/util/kubeclient"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	namespacedSecretName = "aso-credential"
+)
+
+type armClient struct {
+	genericClient  *genericarmclient.GenericClient
+	secretData     map[string][]byte
+	credentialFrom types.NamespacedName
+}
+
+func (c armClient) GenericClient() *genericarmclient.GenericClient {
+	return c.genericClient
+}
+
+func (c armClient) CredentialFrom() types.NamespacedName {
+	return c.credentialFrom
+}
+
+type armClients struct {
```

Comment:
minor: Consider calling this `armClientsCache` (or maybe `armClientCache`, not sure if the `s` is better or not).